token-metadata/program/src/processor/burn/burn_nft.rs

```diff
@@ -35,43 +27,32 @@ pub fn process_burn_nft(program_id: &Pubkey, accounts: &[AccountInfo]) -> Progra
     let spl_token_program_info = next_account_info(account_info_iter)?;
 
     let collection_nft_provided = accounts.len() == 7;
+    let collection_metadata_info = if collection_nft_provided {
+        Some(next_account_info(account_info_iter)?)
+    } else {
+        None
+    };
 
-    let metadata = Metadata::from_account_info(metadata_info)?;
-
-    // If the NFT is a verified part of a collection but the user has not provided the collection
-    // metadata account, we cannot burn it because we need to check if we need to decrement the collection size.
-    if !collection_nft_provided
-        && metadata.collection.is_some()
-        && metadata.collection.as_ref().unwrap().verified
-    {
-        return Err(MetadataError::MissingCollectionMetadata.into());
-    }
-
-    // Ensure this is a Master Edition and not a Print.
-
-    // Scope this so the borrow gets dropped and doesn't conflict with the mut borrow
-    // later in the handler when overwriting data.
-    {
-        let edition_account_data = edition_info.try_borrow_data()?;
+    // Validate accounts
 
-        // First byte is the object key.
-        let key = edition_account_data
-            .first()
-            .ok_or(MetadataError::InvalidMasterEdition)?;
-        if *key != Key::MasterEditionV1 as u8 && *key != Key::MasterEditionV2 as u8 {
-            return Err(MetadataError::NotAMasterEdition.into());
-        }
+    // Assert signer
+    assert_signer(owner_info)?;
 
-        // Next eight bytes are the supply, which must be converted to a u64.
-        let supply_bytes = array_ref![edition_account_data, 1, 8];
-        let supply = u64::from_le_bytes(*supply_bytes);
+    // Assert program ownership.
+    assert_owned_by(metadata_info, program_id)?;
+    assert_owned_by(edition_info, program_id)?;
+    assert_owned_by(mint_info, &spl_token::ID)?;
+    assert_owned_by(token_info, &spl_token::ID)?;
 
-        // Cannot burn Master Editions with existing prints in this handler.
-        if supply > 0 {
-            return Err(MetadataError::MasterEditionHasPrints.into());
-        }
+    // Check program IDs.
+    if spl_token_program_info.key != &spl_token::ID {
+        return Err(ProgramError::IncorrectProgramId);
     }
 
+    // Deserialize accounts.
+    let metadata = Metadata::from_account_info(metadata_info)?;
+    let _: TokenAccount = assert_initialized(token_info)?;
```

Comment:
This is kind of an unused duplicate because it is also done in `assert_currently_holding`.  Maybe  you don't need it here because you are still using `assert_currently_holding`?  In my comment for the new handler I recommended deprecating `assert_currently_holding` since you are following a newer verification approach, but maybe for the old handlers you want to keep it?  The only reason I'm saying anything is because its an extra deserialize.

Reply:
You're right that this is unnecessary and can be safely removed. I think a larger refactor of this legacy handler is out of the scope of the Burn PR tho.